### PR TITLE
Annotation support

### PR DESCRIPTION
**Pull request by atomic77**
Provide support for annotations in faasd with namespaced container
labels

Signed-off-by: Alex Tomic <atomic777@gmail.com>

## Motivation and Context

Annotations are supported in other flavours of openfaas

https://github.com/openfaas/faasd/issues/51
- [ ] I have raised an issue to propose this change **this is required**

## How Has This Been Tested?

I have provided new unit tests and ran an end to end test with faasd deployed to a multipass managed VM, with the following simple configuration:
```yaml
version: 1.0                         
provider:                            
  name: openfaas                     
  gateway: http://127.0.0.1:8080     
functions:                           
  figlet-test:                       
    lang: dockerfile                 
    image: functions/figlet:0.13.0   
    labels:                          
      lab1: "val1"                   
    annotations:                     
      foo: "bar"                     
      qwer: "ty"                     
```
Doing a GET to `/system/function/figlet-test` comes back with the annotation:

```json
{
  "name": "figlet-test",
  "image": "",
  "invocationCount": 0,
  "replicas": 1,
  "envProcess": "",
  "availableReplicas": 1,
  "labels": {
    "lab1": "val1"
  },
  "annotations": {
    "foo": "bar",
    "qwer": "ty"
  },
  "namespace": "openfaas-fn"
}
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Commits:

- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] My commit message has a body and describe how this was tested and why it is required.
- [x] I have signed-off my commits with `git commit -s` for the Developer Certificate of Origin (DCO)

Code:

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.

Docs:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by alexellis** [comment]
I'm not seeing any immediate reason as to why CI is failing. 

I restarted the build on Travis and have seen the same error at the same point, twice:

```
/usr/local/bin/faas-cli list -v
497Function                      	Image                                   	Invocations    	Replicas
498figlet                        	docker.io/functions/figlet:0.13.0       	0              	1    
499/usr/local/bin/faas-cli describe figlet | grep testing
500Labels:              testing : true
501uname | /usr/local/bin/faas-cli invoke figlet
502Server returned unexpected status code: 404 - error finding function figlet.: Get http://faasd-provider:8081/system/function/figlet?namespace=: dial tcp: i/o timeout
503Makefile:43: recipe for target 'test-e2e' failed
504make: *** [test-e2e] Error 1
505The command "make test-e2e" exited with 2.
```

Alex, maybe you can try the steps from the e2e tests on your own local machine with the build in place?

**Comment by alexellis** [comment]
@LucasRoesler please could you review before this gets merged?

**Comment by alexellis** [comment]
Let's just make sure we don't run into this issue too - https://github.com/openfaas/faasd/pull/102

File: pkg/provider/handlers/deploy_test.go
Comment (alexellis) [inline pkg/provider/handlers/deploy_test.go:10]: I would like a few more unit tests if possible?

no annotations + no labels
no annotations but labels
annotations  but no labels

Would it make sense to use a test-table pattern?

https://blog.alexellis.io/golang-writing-unit-tests/

File: pkg/provider/handlers/deploy_test.go
Comment (LucasRoesler) [inline pkg/provider/handlers/deploy_test.go:13]: This can be improved by also adding a name to each test case and using a subtest

```go
    // Test each combination of nil/non-nil annotation + label
	tables := []struct {
        name string
		label      map[string]string
		annotation map[string]string
		result     map[string]string
	}{
		{"empty annotations returns empty lable map", nil, nil, map[string]string{}},
		{"...",map[string]string{"L1": "V1"}, nil, map[string]string{"L1": "V1"}},
		{"....",nil, map[string]string{"A1": "V2"}, map[string]string{fmt.Sprintf("%sA1", annotationLabelPrefix): "V2"}},
		{
            "....",
			map[string]string{"L1": "V1"}, map[string]string{"A1": "V2"},
			map[string]string{"L1": "V1", fmt.Sprintf("%sA1", annotationLabelPrefix): "V2"},
		},
	}
    for _, pair := range tables {
        t.Run(pair.name, func(t *testing.T){
            request := &types.FunctionDeployment{
                Labels:      &pair.label,
                Annotations: &pair.annotation,
            }
            val, err := buildLabels(request)
            if err != nil {
                t.Fatalf("want: no error got: %v", err)
            }
            if !reflect.DeepEqual(val, pair.result) {
                t.Errorf("Got: %s, expected %s", val, pair.result)
            }
        })
	}
```

Comment (atomic77) [reply pkg/provider/handlers/deploy_test.go:13]: Thanks for the feedback - I've added the test name and included more newlines to make the test table a bit more readable and clear.

**Review by LucasRoesler — APPROVED** [review]

**Comment by LucasRoesler** [comment]
@alexellis looks good with the new tests :shipit:

**Comment by alexellis** [comment]
Merging. Thank you for the review Lucas. This new feature should unblock #119 and the other [event connectors](https://docs.openfaas.com/reference/triggers/)

**Pull request merged**
